src/Propulsion.Kafka/Consumers.fs

```diff
@@ -283,62 +283,150 @@ module Core =
+        /// Processor pumps until `Stop()` is requested.
+        /// Handler `Choice1Of2` result must indicate Index at which next processing will proceed (which can trigger discarding of earlier items on that stream)
+        /// Handler `Choice2Of2` result marks the processing of a stream failed (which will then be offered again for retry purposes on the next cycle)
+        /// <c>keyValueToStreamEvents</c> is often implemented via <c>StreamNameSequenceGenerator.KeyValueToStreamEvent</c>
+        static member Start<'Res>
+            (   log : ILogger, config : KafkaConsumerConfig,
+                /// often implemented via <c>StreamNameSequenceGenerator.KeyValueToStreamEvent</c>
+                keyValueToStreamEvents : KeyValuePair<string, string> -> Propulsion.Streams.StreamEvent<_> seq,
+                handle : StreamName * Streams.StreamSpan<_> -> Async<'Res>, maxDop,
+                stats : Streams.Scheduling.StreamSchedulerStats<OkResult<'Res>, FailResult>,
+                /// Prevent batches being consolidated prior to scheduling in order to maximize granularity of consumer offset updates
+                ?maximizeOffsetWriting,
+                ?pipelineStatsInterval, ?maxSubmissionsPerPartition, ?pumpInterval, ?logExternalState, ?idleDelay, ?maxBatches) =
+            StreamsConsumer.Start<KeyValuePair<string, string>, 'Res>(
+                log, config, Bindings.mapConsumeResult, keyValueToStreamEvents, handle, maxDop, stats,
+                ?pipelineStatsInterval = pipelineStatsInterval,
+                ?maxSubmissionsPerPartition = maxSubmissionsPerPartition,
+                ?pumpInterval = pumpInterval,
+                ?logExternalState = logExternalState,
+                ?idleDelay = idleDelay,
+                ?maxBatches = maxBatches,
+                ?maximizeOffsetWriting = maximizeOffsetWriting)
+
+    // Maps a (potentially `null`) message key to a valid {Category}-{StreamId} StreamName for routing and/or propagation through StreamsProjector
+    let parseMessageKey defaultCategory = function
+        | null -> FsCodec.StreamName.create defaultCategory ""
+        | key -> Propulsion.Streams.StreamName.parseWithDefaultCategory defaultCategory key
+
+/// StreamsConsumer buffers and deduplicates messages from a contiguous stream with each message bearing an `index`.
+/// Where the messages we consume don't have such characteristics, we need to maintain a fake `index` by keeping an int per stream in a dictionary
+type StreamNameSequenceGenerator() =
+
+    // we synthesize a monotonically increasing index to render the deduplication facility inert
+    let indices = System.Collections.Generic.Dictionary()
+
+    /// Generates an index for the specified StreamName
+    member __.GenerateIndex(streamName : StreamName) =
+        let streamName = FsCodec.StreamName.toString streamName
+        match indices.TryGetValue streamName with
+        | true, v -> let x = v + 1 in indices.[streamName] <- x; int64 x
```

Comment:
Can you store `int64` in the dictionary so you don't need any casting?

`let x = v + 1L in indices.[streamName] <- x; x`

Reply:
good spot - should have got you to review this before!